NetKet/GroundState/exact_diagonalization.hpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2018 Alexander Wietek - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_EXACT_DIAGONALIZATION_HPP
+#define NETKET_EXACT_DIAGONALIZATION_HPP
+
+#include <vector>
+
+#include <ietl/randomgenerator.h>
+#include <ietl/lanczos.h>
+
+#include "Hamiltonian/MatrixWrapper/matrix_wrapper.hpp"
+
+namespace netket {
+
+  std::vector<double> eigenvalues_lanczos(const Hamiltonian &hamiltonian,
+                                          int first_n = 1) {
+    using complex = std::complex<double>;
+    using normal_dist_t = std::uniform_real_distribution<double>;
+    using matrix_t = Eigen::SparseMatrix<complex>;
+    using vectorspace_t = ietl::vectorspace<complex>;
+    using iter_t = ietl::lanczos_iteration_nlowest<double>;
+    using lanczos_t = ietl::lanczos<matrix_t, vectorspace_t>;
+
+    // TODO: make these parameters accessible to user
+    int max_iter = 1000;
+    int seed = 42;
+    double precision = 1e-8;
+
+    SparseMatrixWrapper<Hamiltonian> matrix(hamiltonian);
+    const matrix_t numerical_matrix = matrix.GetMatrix();
```

Comment:
This copies the whole sparse matrix instead of taking a reference to the one store in `matrix`, so at least this will take twice as much memory as necessary (since the matrix wrapper is still in scope).